### PR TITLE
added lint rules to check for copyright header in .java files

### LINKED ISSUE #8211: Add Lint check to ensure all .java files have Copyright Headers
With more new contributors, we should automate ensuring that each new file has a copyright header, rather than leaving this to code review.

We can do this via a lint check. For example: https://github.com/ankidroid/Anki-Android/blob/db82ab008a90ff89e77c5c5637b2e33be17ab04a/lint-rules/src/main/java/com/ichi2/anki/lint/rules/InconsistentAnnotationUsage.java

It would be ideal if the lint check guided users to the correct setting screen in Android Studio to set up the copyright:

This can be done via an explanation: https://github.com/ankidroid/Anki-Android/blob/db82ab008a90ff89e77c5c5637b2e33be17ab04a/lint-rules/src/main/java/com/ichi2/anki/lint/rules/InconsistentAnnotationUsage.java#L37-L39

image

It may also be useful to provide users with a template:

```
 Copyright (c) $today.year Your Name Here <example@example.com>

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 3 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT ANY
 WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 PARTICULAR PURPOSE. See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>.
```

### PR DESCRIPTION
**Pull request by almas33**
## Pull Request template
added a lint rule which will check that all files have copyright header and if not will explain contributer how to do so
## Fixes
Fixes #8211 

## Approach
Feed `Copyright Header` in `Pattern`
`Matcher` will tell if that pattern is found
Reference use : https://github.com/vanniktech/lint-rules/blob/master/lint-rules-android-lint/src/main/java/com/vanniktech/lintrules/android/InvalidSingleLineCommentDetector.kt
## How Has This Been Tested?

CI should tell

## Learning (optional, can help others)
https://github.com/vanniktech/lint-rules
## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by almas33** [comment]
Sorry, but Not sure if this will work as expected waiting for checks...

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    public List<Class<? extends UElement>> getApplicableUastTypes() {
+        return Collections.singletonList(UClass.class);
+    }
+
+
+    @Override
+    public void afterCheckFile(@NotNull Context context) {
+        String source = context.getContents().toString();
+        Pattern pattern = Pattern.compile("/*\n" +
+                " *  Copyright (c) $today.year David Allison <davidallisongithub@gmail.com>\n" +
```
Comment (david-allison-1) [inline lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java:73]: This needs to be less restrictive, there are many comment styles within the codebase. 

Especially don't hardcode my name or the year, I don't want to have to make every file 😅

File: lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            " *  You should have received a copy of the GNU General Public License along with\n" +
+            " *  this program.  If not, see <http://www.gnu.org/licenses/>.\n" +
+            " */";
+
+
+
+
+    @Test
+    public void showsErrorForNoCopyrightHeader() {
+        lint()
```
Comment (david-allison-1) [inline lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java:50]: Surely this file should succeed, and a file without the copyright header should fail?

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class CopyrightHeaderChecker extends Detector implements SourceCodeScanner {
+    @VisibleForTesting
+    static final String ID = "CopyrightHeaderChecker";
+    @VisibleForTesting
+    static final String DESCRIPTION = "Add Copyright Header in each file";
+    private static final String EXPLANATION = "All files in AnkiDroid must contain  Copyright Header. " +
```
Comment (david-allison-1) [inline lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java:43]: You might want to add newlines here as well.

```suggestion
    private static final String EXPLANATION = "All files in AnkiDroid must contain a GPLv3-compatible copyright header" +
```

You should also provide a link to a "default" copyright header that someone can copy/paste (maybe a wiki page?).

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            "The copyright header can be set in Settings - Editor - Copyright - Copyright Profiles - Add Profile - AnkiDroid. " +
+            "Search in Settings for 'Copyright'";
+    private static final Implementation implementation = new Implementation(CopyrightHeaderChecker.class, Scope.JAVA_FILE_SCOPE);
+    public static final Issue ISSUE = Issue.create(
+            ID,
+            DESCRIPTION,
+            EXPLANATION,
+            Constants.ANKI_TIME_CATEGORY,
```
Comment (david-allison-1) [inline lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java:51]: Might want to change these - they're fairly unimportant for now

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    @Nullable
+    @Override
+    public List<Class<? extends UElement>> getApplicableUastTypes() {
+        return Collections.singletonList(UClass.class);
+    }
+
+
+    @Override
+    public void afterCheckFile(@NotNull Context context) {
+        String source = context.getContents().toString();
```
Comment (david-allison-1) [inline lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java:71]: Unsure here: we may be able to get a speedup if we don't use `.toString()`

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Seeing this made my day, thank you! 

Things like these mean that code reviews will take less time to do, and we have more time to make AnkiDroid better

**Comment by david-allison-1** [comment]
You can use `gradlew lintRelease` to test your code on the codebase locally.

I expect that it'll fail - but not crash, we'll likely have a few files which are missing copyright, which we'll deal with

**Comment by david-allison-1** [comment]
@almas33 Do you expect you'll have time to work on feedback?

**Comment by almas33** [comment]
Yes definitely, Sorry I got busy due to college projects

**Comment by david-allison-1** [comment]
> Yes definitely, Sorry I got busy due to college projects

No worries!

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java (hunk 4 above)
Comment (almas33) [reply lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java:51]: But if I remove them arguements will be less

**Comment by almas33** [comment]
> You can use `gradlew lintRelease` to test your code on the codebase locally.
> 
> I expect that it'll fail - but not crash, we'll likely have a few files which are missing copyright, which we'll deal with. Once you get to this stage, let's make it public so we can discuss what to do with the files

It is getting successful that means something is wrong

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java (hunk 4 above)
Comment (david-allison-1) [reply lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java:51]: rebase onto `master`, then use `Constants.ANKI_CODE_STYLE_CATEGORY`

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
+/*
+ *  Copyright (c) 2021 David Allison <davidallisongithub@gmail.com>
```
Comment (david-allison-1) [inline lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java:2]: Not my file

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+
+    @Override
+    public void afterCheckFile(@NotNull Context context) {
+        CharSequence source = context.getContents();
+        Pattern pattern = Pattern.compile("/*\n" +
+                " *  This program is free software; you can redistribute it and/or modify it under\n" +
+                " *  the terms of the GNU General Public License as published by the Free Software\n" +
+                " *  Foundation; either version 3 of the License, or (at your option) any later\n" +
```
Comment (david-allison-1) [inline lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java:79]: Just checking for this fragment should be enough
```
                " * either version 3 of the License, or (at your option) any later\n" +
```

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                " *\n" +
+                " *  You should have received a copy of the GNU General Public License along with\n" +
+                " *  this program.  If not, see <http://www.gnu.org/licenses/>.\n" +
+                " */");
+        Matcher matcher = pattern.matcher(source);
+        boolean found = false;
+        if (matcher.find()) {
+            found = true;
+        }
+        if (!found) {
```
Comment (david-allison-1) [inline lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java:94]:
```suggestion
        if (!pattern.matcher(source).find()) {
```

File: lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
+/*
+ *  Copyright (c) 2021 David Allison <davidallisongithub@gmail.com>
```
Comment (david-allison-1) [inline lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java:2]: Not me

File: lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+            " *  version.\n" +
+            " *\n" +
+            " *  This program is distributed in the hope that it will be useful, but WITHOUT ANY\n" +
+            " *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A\n" +
+            " *  PARTICULAR PURPOSE. See the GNU General Public License for more details.\n" +
+            " *\n" +
+            " *  You should have received a copy of the GNU General Public License along with\n" +
+            " *  this program.  If not, see <http://www.gnu.org/licenses/>.\n" +
+            " */";
+
```
Comment (david-allison-1) [inline lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java:42]: Also test for a failure

File: lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    @Test
+    public void fileWithCopyrightHeaderPasses() {
+        lint()
+                .allowMissingSdk()
+                .allowCompilationErrors()
+                .files(create(mCopyrightHeader))
+                .issues(CopyrightHeaderChecker.ISSUE)
+                .run()
+                .expectErrorCount(1)
```
Comment (david-allison-1) [inline lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java:54]: This should be fine?

**Review by david-allison-1 — CHANGES_REQUESTED** [review]

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    @Override
+    public List<Class<? extends UElement>> getApplicableUastTypes() {
+        return Collections.singletonList(UClass.class);
+    }
+
+
+    @Override
+    public void afterCheckFile(@NotNull Context context) {
+        CharSequence source = context.getContents();
+        Pattern pattern = Pattern.compile("/*\n" +
```
Comment (david-allison-1) [inline lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderChecker.java:76]: Compile the pattern in a static for performancee

File: lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java (hunk 11 above)
Comment (almas33) [reply lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java:54]: Got little confused from this and comment from previous commit , please clarify
we have to create 2 tests :-
1. filewithCopyrightHeaderfails
    expectErrorCount(0)
2. showsNoErrorOnMissingCopyrightHeader
   expectErrorCount(0)

Comment (david-allison-1) [reply lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java:54]: A file with a copyright header does not fail the lint check

A file with no copyright header fails the lint check (1 error)

Comment (almas33) [reply lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java:54]: Yes , So in first commit name of test was `showsErrorForNoCopyrightHeader() `  (1 error)
but you said 

> Surely this file should succeed, and a file without the copyright header should fail?

So, I think it was doing the same thing before
Sorry, If I'm not understanding properly

In the last point, I was asking about the use of  `String mCopyrightHeader`  used in line `51` and does it have to be same as `Pattern`

Comment (david-allison-1) [reply lint-rules/src/test/java/com/ichi2/anki/lint/rules/CopyrightHeaderCheckerTest.java:54]: As far as I recall, a file with a copyright header was passed into `showsErrorForNoCopyrightHeader` in that instance, sorry for being unclear (or incorrect)

----

We should have an example of a full, correct copyright header, no use in checking for the search snippet

If our code misses an example which exists in the codebase, we should add these as test cases, and fix the issue to protect against regressions (for example: the regex may need better newline handling)

**Comment by david-allison-1** [comment]
Fixed in #8662

**Pull request closed without merging**